### ISSUE TITLE
Having 4 different Docker images with different ways of being configured is repetitive and confusing

### ISSUE THREAD

**Issue by reivilibre** [matrix-org/synapse#12847]
We have:

* 'the image' as published on Docker Hub (this one is user-facing and we don't want to break the interface)
* a workers variant (I don't know who uses this other than Complement..?)
* a Complement variant (SQLite only!)
* a Complement + Postgres + workers variant

They don't even configure themselves in the same way. For example, one of them uses ugly `sed` incantations to configure appservices, the other two use slightly different Jinja2 (though they also have good reason for being slightly different, it still seems like they could be unified).

It seems like, as a first step to restore some sanity, we could for example remove the Complement+SQLite image and make it a special case of the workers one?
This would also open up the doors to having Complement+Postgres (without workers).

I will have a look at this to try and see about unifying them. I will probably leave the user-facing image alone.

New idea?:
* 'the image'
* Workers + Postgres
    * Complement + Workers + Postgres
        * Complement + monolith + Postgres (or can use SQLite via env var?); special case of 'no workers'

**Comment by richvdh** [comment]
> a workers variant (I don't know who uses this other than Complement..?)

ISTR @MatMaul may have used it for something?

**Comment by richvdh** [comment]
> we could for example remove the Complement+SQLite image

One thing about this is that, because the Complement+SQLite image is a pretty light modification to the base Synapse image, it gives some confidence that our base image starts and runs, which isn't necessarily true for the full-on Complement+Postgres+workers image.

That said, I suspect that's a fairly illusory safety net, and if that's something we actually want to test, we should do it a different way.

In short: your idea sounds good.

**Comment by DMRobertson** [comment]
> * 'the image'

Unclear how Pyston will fit into this: see also #12850

**Comment by reivilibre** [comment]
> > * 'the image'
> 
> Unclear how Pyston will fit into this: see also #12850

That's a good point. I wonder how tricky it'd be to make this a build argument.

One thing I've used recently is https://earthly.dev/ (I think Dan mentioned it a long time ago). It's kind of a build tool that uses Docker as its way to do things.

**Comment by DMRobertson** [comment]
> I wonder how tricky it'd be to make Pyston a build argument.

I'd rather see us ship only one interpreter in our docker images. We have enough on our plate as-is; no reason to double our state space.

**Comment by reivilibre** [comment]
> I'd rather see us ship only one interpreter in our docker images. We have enough on our plate as-is; no reason to double our state space.

That's certainly one option, but does this mean that we'll have to choose between running Complement tests against Pyston (which we will maybe be using on m.org) or against CPython (which I think we would be somewhat unwilling to drop support for)?

I suppose you're saying 'ship'; I certainly have no objection to having only one image we dish out to admins, but it's still useful for development/testing tasks to be able to customise it a bit. I don't know how extensive the changes are to make our Dockerfile compatible with Pyston, but if it's not too extensive, I still think it'd be best to keep it as an argument.

**Comment by anoadragon453** [comment]
Switching our Dockerfile to Pyston immediately *will* break some images that are based on it, as admins will have inevitably been relying on binary wheels which aren't compiled for Pyston - meaning they'll need to install additional native dependencies in their derived Dockerfile. See where we ran into this problem when `pip install`ing poetry in our own dockerfile: https://github.com/matrix-org/synapse/pull/12850#discussion_r880334811

This isn't a long-term issue - it just requires an announcement and transition period from CPython to Pyston. For now, I'd suggest we make a `-pyston` tag to make our release and latest tags which the community can test and send feedback for. Doing so is a simple modification to our current deployment yaml: https://github.com/matrix-org/synapse/pull/12850/files#diff-3f5366f6d6df3ec1179e5efadc6f350bfa88eebf4e2da589b4d94ccb85ae5e94R16-R23

For our own testing, there *ideally shouldn't* be a difference between the behaviour of CPython and Pyston, but there's always a chance of an implementation bug. Unless we want to double the number of GHA workers(?), I suggest we test on whichever we consider the default interpreter in the docker images (CPython for now). This would mean that what we're testing is slightly different from what we're running on matrix.org (Pyston).

As an aside: if we want to try running the Complement worker tests with Pyston once #12850 is merged to see if it runs faster, that would be interesting.

**Comment by reivilibre** [comment]
> As an aside: if we want to try running the Complement worker tests with Pyston once https://github.com/matrix-org/synapse/pull/12850 is merged to see if it runs faster, that would be interesting.

I would be interested, but I suspect the time to build the image will blow up way more than the time it would save

**Comment by anoadragon453** [comment]
> I would be interested, but I suspect the time to build the image will blow up way more than the time it would save

Hmm, could we be smarter about how we build our Complement docker images, and install the PR's Synapse on top of the latest released Synapse (pyston) image. Therefore bypassing the long wait times for cryptography and other dependencies to compile and install?

**Comment by reivilibre** [comment]
> Hmm, could we be smarter about how we build our Complement docker images, and install the PR's Synapse on top of the latest released Synapse (pyston) image. Therefore bypassing the long wait times for cryptography and other dependencies to compile and install?

Yes, potentially, though this means we need to publish an image somewhere (which was a controversial point)

**Comment by anoadragon453** [comment]
Yeah, we can do it after adding Pyston to the CI I suppose.
